server/src/main/java/org/opensearch/index/engine/NRTReplicationEngine.java

```diff
@@ -0,0 +1,468 @@
+import org.opensearch.index.seqno.LocalCheckpointTracker;
+import org.opensearch.index.seqno.SeqNoStats;
+import org.opensearch.index.seqno.SequenceNumbers;
+import org.opensearch.index.translog.DefaultTranslogDeletionPolicy;
+import org.opensearch.index.translog.Translog;
+import org.opensearch.index.translog.TranslogConfig;
+import org.opensearch.index.translog.TranslogDeletionPolicy;
+import org.opensearch.index.translog.TranslogStats;
+import org.opensearch.search.suggest.completion.CompletionStats;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.CountDownLatch;
+import java.util.function.BiFunction;
+import java.util.function.LongConsumer;
+import java.util.function.LongSupplier;
+import java.util.stream.Stream;
+
+/**
+ * This is an {@link Engine} implementation intended for replica shards when Segment Replication
+ * is enabled.  This Engine does not create an IndexWriter, rather it refreshes a {@link NRTReplicationReaderManager}
+ * with new Segments when received from an external source.
+ *
+ * @opensearch.internal
+ */
+public class NRTReplicationEngine extends Engine {
+
+    private volatile SegmentInfos lastCommittedSegmentInfos;
+    private final NRTReplicationReaderManager readerManager;
+    private final CompletionStatsCache completionStatsCache;
+    private final LocalCheckpointTracker localCheckpointTracker;
+    private final Translog translog;
+
+    public NRTReplicationEngine(EngineConfig engineConfig) {
+        super(engineConfig);
+        store.incRef();
+        NRTReplicationReaderManager readerManager = null;
+        try {
+            lastCommittedSegmentInfos = store.readLastCommittedSegmentsInfo();
+            readerManager = new NRTReplicationReaderManager(OpenSearchDirectoryReader.wrap(getDirectoryReader(), shardId));
+            final SequenceNumbers.CommitInfo commitInfo = SequenceNumbers.loadSeqNoInfoFromLuceneCommit(
+                this.lastCommittedSegmentInfos.getUserData().entrySet()
+            );
+            this.localCheckpointTracker = new LocalCheckpointTracker(commitInfo.maxSeqNo, commitInfo.localCheckpoint);
+            this.completionStatsCache = new CompletionStatsCache(() -> acquireSearcher("completion_stats"));
+            this.readerManager = readerManager;
+            this.readerManager.addListener(completionStatsCache);
+            this.translog = openTranslog(
+                engineConfig,
+                getTranslogDeletionPolicy(engineConfig),
+                engineConfig.getGlobalCheckpointSupplier(),
+                localCheckpointTracker::markSeqNoAsPersisted
+            );
+        } catch (IOException e) {
+            IOUtils.closeWhileHandlingException(store::decRef, readerManager);
+            throw new EngineCreationFailureException(shardId, "failed to create engine", e);
+        }
+    }
```

Comment:
We should always try release any releasable within the same code block

Reply:
Are you referring to store.incRef() ? This is similar to InternalEngine that holds that ref on the class and releases in `closeNoLock`.

Reply:
@Bukhtawar could you pls clarify here which releasable you are referring to in this constructor?

Reply:
That's right I meant the store ref